modAionImpl/src/org/aion/zero/impl/db/PendingBlockStore.java

```diff
@@ -0,0 +1,816 @@
+     * the needed height and the parent block gets imported. Is used by the functionality receiving
+     * status blocks.
+     *
+     * @implNote The status blocks received impact the functionality of the base value generation
+     *     {@link #nextBase(long, long)} for requesting blocks ahead of import time.
+     */
+    public boolean addStatusBlock(AionBlock block) {
+
+        // nothing to do with null parameter
+        if (block == null) {
+            return false;
+        }
+
+        lock.writeLock().lock();
+
+        try {
+            // skip if already stored
+            if (!indexSource.get(block.getHash()).isPresent()) {
+
+                // find parent queue hash
+                Optional<byte[]> existingQueueHash = indexSource.get(block.getParentHash());
+                byte[] currentQueueHash = null;
+                List<AionBlock> currentQueue = null;
+
+                // get existing queue if present
+                if (existingQueueHash.isPresent()) {
+                    // using parent queue hash
+                    currentQueueHash = existingQueueHash.get();
+
+                    // append block to queue
+                    currentQueue = queueSource.get(currentQueueHash);
+                } // do not add else here!
+
+                // when no queue exists OR problem with existing queue
+                if (currentQueue == null || currentQueue.size() == 0) {
+                    // start new queue
+
+                    // queue hash = the node hash
+                    currentQueueHash = block.getHash();
+                    currentQueue = new ArrayList<>();
+
+                    // add (to) level
+                    byte[] levelKey = ByteUtil.longToBytes(block.getNumber());
+                    List<byte[]> levelData = levelSource.get(levelKey);
+
+                    if (levelData == null) {
+                        levelData = new ArrayList<>();
+                    }
+
+                    levelData.add(currentQueueHash);
+                    levelSource.put(levelKey, levelData);
+                }
+
+                // NOTE: at this point the currentQueueHash was initialized
+                // either with a previous hash OR the block hash
+
+                // index block with queue hash
+                indexSource.put(block.getHash(), currentQueueHash);
+
+                // add element to queue
+                currentQueue.add(block);
+                queueSource.put(currentQueueHash, currentQueue);
+
+                // update status tracking
+                ByteArrayWrapper hash = ByteArrayWrapper.wrap(currentQueueHash);
+                QueueInfo info = status.get(hash);
+                if (info == null) {
+                    if (Arrays.equals(currentQueueHash, block.getHash())) {
+                        info =
+                                new QueueInfo(
+                                        currentQueueHash, block.getNumber(), block.getNumber());
```

Comment:
why the constructor has two same args input?

Reply:
The two parameters had different meanings, the first and last number in the queue. Also, I couldn't use two constructors with the same parameters. Refactored in 153fa7c.